### ISSUE TITLE
fmn.lib:'set' object has no attribute '__getitem__'

### ISSUE THREAD

**Issue by jeremycline** [fedora-infra/fmn#194]
It looks like one API was expecting a dictionary and is now getting a set, causing it to crash when running the ``user_package_filter`` rule:

```
May 17 02:02:56 notifs-backend01.phx2.fedoraproject.org python2[12137]: ERROR:fmn.lib:'set' object has no attribute '__getitem__'
May 17 02:02:56 notifs-backend01.phx2.fedoraproject.org python2[12137]: Traceback (most recent call last):
May 17 02:02:56 notifs-backend01.phx2.fedoraproject.org python2[12137]: File "/usr/lib/python2.7/site-packages/fmn/lib/__init__.py", line 84, in matches
May 17 02:02:56 notifs-backend01.phx2.fedoraproject.org python2[12137]: value = fn(config, message, **arguments)
May 17 02:02:56 notifs-backend01.phx2.fedoraproject.org python2[12137]: File "/usr/lib/python2.7/site-packages/fmn/rules/generic.py", line 107, in user_package_filter
May 17 02:02:56 notifs-backend01.phx2.fedoraproject.org python2[12137]: return _user_package_intersection(fasnick, message, config, all_acls)
May 17 02:02:56 notifs-backend01.phx2.fedoraproject.org python2[12137]: File "/usr/lib/python2.7/site-packages/fmn/rules/generic.py", line 67, in _user_package_intersection
May 17 02:02:56 notifs-backend01.phx2.fedoraproject.org python2[12137]: config, fasnick, acls)
May 17 02:02:56 notifs-backend01.phx2.fedoraproject.org python2[12137]: File "/usr/lib/python2.7/site-packages/fmn/rules/utils.py", line 148, in get_packages_of_user
May 17 02:02:56 notifs-backend01.phx2.fedoraproject.org python2[12137]: packages[namespace].update(subset[namespace])
May 17 02:02:56 notifs-backend01.phx2.fedoraproject.org python2[12137]: TypeError: 'set' object has no attribute '__getitem__'
```

**Comment by jeremycline** [comment]
Okay, after a lot of head-scratching I've tracked this issue down. The problem turned out to be a caching problem. The backend runs an instance of Redis, which is (apparently by default) set to persist to disk. In addition, dogpile.cache appears to be mis-configured or has a bug where it's not setting the TTL on redis keys, leading to an ever-growing, persisted "cache" with tons of stale objects.

I'm leaving this open until we find out if we're mis-configuring dogpile.cache or whether it's a bug.

**Comment by jeremycline** [comment]
Production runs with dogpile.cache version 0.5.7, which suffers from [a bug fixed in 0.6.0](https://dogpilecache.readthedocs.io/en/latest/changelog.html#change-0.6.0-bug). We configure the cache using the dictionary from ``config['fmn.rules.cache']``. However, two other caches also use this configuration (fmn.consumer.fmn_fasshim and fmn.consumer.worker) and because it's pre-0.6.0, the arguments in the configuration dictionary are popped:

```
>>> config = load_config()
>>> config['fmn.rules.cache']
{'arguments': {'redis_expiration_time': 7200, 'host': 'localhost', 'distributed_lock': True, 'db': 0, 'port': 6379}, 'backend': 'dogpile.cache.redis'}
>>> utils._cache.configure(**config['fmn.rules.cache'])
<dogpile.cache.region.CacheRegion object at 0x2167390>
>>> config['fmn.rules.cache']
{'arguments': {'distributed_lock': True}, 'backend': 'dogpile.cache.redis'}
```

This results in a race condition where whoever configures their cache first sets their keys to expire, and the rest don't and therefore the keys never expire. Combine this with the redis default to persist to disk and you get a database instead of a cache. 😢.
